### ISSUE TITLE
Ability to change sync URL, username, passwords.

### ISSUE THREAD

**Issue by ujwalSussol** [openmsupply/mobile#1504]
## Is your feature request related to a problem? Please describe.
Ability to change sync URL, username, passwords.

The frustration : Modifying the sync password on the server means that the mobile tablet is totally disabled and sync will not work. To get sync to work, the only thing that can be done is to revert the server to the old password.  @raviSussol has had a tough time supporting Ivory Coast after changing the password on the cloud server.

## Describe the solution you'd like

Similar to mSupply desktop, we need to have the ability to modify sync setting on the mobile tablet such as : 

- URL
- Sync password

We prevent the : Sync user name  from being changed for security reasons.

## Describe alternatives you've considered

N/A

## Additional context
This was considered in the past :  #347  but was ignored. This shouldn't be ignored now as it is a problem the support guys are facing.

Do this and I have an an idea which will help us to import a store data into mobile :-)  ....  let get this done first.

**Comment by GaryWilletts** [comment]
Does this need a corresponding issue on the mobile side too?

**Comment by joshxg** [comment]
Main discussion point for implementation is that changing the sync URL and sync password will break the app if done incorrectly.
- Ensure this can be done from the first page and login page.
- I think it is time for a settings page. Shift realm browser and export data there. Possibly store custom data editing in the future.
- Need some security layer - password sent to access settings during initial dump? Then, do we need a way to change the password to access settings? :)

**Comment by Chris-Petty** [comment]
@ujwalSussol I think this would be useful generally but I think it's approaching your reported problem from the wrong side. If you change the sync password and lose track of it, then rely on updating mobile to use a different one, that means having solid remote or physical access to the device, or a local power user with the knowhow and clear instructions. That doesn't seem like an easy or reliable alternative.

I think a far simpler way, in both implementation and administration, would be to have a way to disable sync for a site centrally that doesn't affect the password. A boolean `isActive` (or rather `isSyncPaused`, we still want records to be generated!) and a checkbox, with a condition check added to incoming sync requests?

All of that said, why were passwords being changed? I thought we only did that to avoid accidentally doubling up devices on a sync site, but the better solution to that is the hardware UUID check (though that has some issues to work through still: #1449 #1290).

---

If we do this regardless to any of the above I agree with @joshxg, this would be good but needs some security.

> - Ensure this can be done from the first page and login page.

Fair idea, otherwise you are hosed if you haven't logged in at least once (or don't know any saved user details). Final point might play against this.

> - I think it is time for a settings page. Shift realm browser and export data there. Possibly store custom data editing in the future.

Agree. Failing my dreams of restructuring mobile main page all together, I think a small cog button/menu should be somewhere on screen.

> - Need some security layer - password sent to access settings during initial dump? Then, do we need a way to change the password to access settings? :)

More elaborate idea... Should be a user pref. Actually sync users. Users that have "Mobile device administration" edit settings. Should aim for at least one admin user with login rights to mobile stores in a system. Can't be "sussol" as that is hardcoded out of login through REST API.

---

> Does this need a corresponding issue on the mobile side too?

@GaryWilletts do you mean server side?

**Comment by ujwalSussol** [comment]
> Does this need a corresponding issue on the mobile side too?

@GaryWilletts    : this issue is for the mobile tablet. mSupply desktop allows the password to be modified on mSupply cloud and as well as Single user mSupply desktop.

**Comment by ujwalSussol** [comment]
@Chris-Petty   :  being able to change sync password increases security.  If the old password is known by a ex-staff then we may want to change to a new password... Currently we have no way to feed the updated password to the mobile. 

@Chris-Petty   You know that Last pass isn't totally reliable...  so what you think you saved is not saved......   So would be great to be able to set password to a new one. 

Generally being able to change password increases security.   At this moment, password can only be changed during initial setup of the mobile. 

I imagined a Setting page to modify URL and password if needed.  This should cause any harm. the only thing is sync won't work if the settings are wrong. 

SureMdm allows us to update the password if needed.

**Comment by GaryWilletts** [comment]
> @GaryWilletts do you mean server side?
@Chris-Petty I could have sworn this was an mSupply issue not mobile when I read it first! Yes, I mean an issue in the other repo too i.e. mSupply (server side). I guess we need to decide exactly what the solution is then create the issue in mSupply. Unless there doesn't need to be one...

@ujwalSussol I know mSupply allows the changing of the password but it doesn't transmit it anywhere. I was thinking there would probably need to be corresponding changes in mSupply for the solution to this issue.

**Comment by raviSussol** [comment]
>@ujwalSussol I think this would be useful generally but I think it's approaching your reported problem from the wrong side. If you change the sync password and lose track of it, then rely on updating mobile to use a different one, that means having solid remote or physical access to the device, or a local power user with the knowhow and clear instructions. That doesn't seem like an easy or reliable alternative.

@Chris-Petty this would be the same case as for the desktop satellite site. We use to control and configure all sync related settings by ourselves either remotely or physically. And I also thought that some mSupply settings like this (i.e. sync) should always be done by us (Sussol members) either in mobile or in desktop. Yesterday I had to revert back the sync password for the `CSRNGATTAKRO` to the old one and re-synced the tablet's data to the cloud server. And then again change back to a new password in the cloud server and flushed data back in the tablet and re-initialized the sync setup in it. Luckily we had a backup of old sync passwords otherwise it would have been a big hassle as they were more than 500 sync out records in that tablet to be synced to the cloud server.

>I think a far simpler way, in both implementation and administration, would be to have a way to disable sync for a site centrally that doesn't affect the password. A boolean isActive (or rather isSyncPaused, we still want records to be generated!) and a checkbox, with a condition check added to incoming sync requests?

I like the idea that to handle above case centrally. It would be nice if we can control mobile incoming data as per our need centrally regardless of having with/without sync password. But this still requires a high alert monitoring in mobile site by us as we need to check the mobile site time to time that the partial sync is completed or not from the mobile to the cloud server and then turn off the checkbox in the cloud server to stop unauthorized data from the mobile which takes us back to the above case again that a member from Sussol need to monitor all these tasks. Otherwise we would end up having accidental data syncing from the multiple devices for one site if we let customer to do it.

>All of that said, why were passwords being changed? I thought we only did that to avoid accidentally doubling up devices on a sync site, but the better solution to that is the hardware UUID check (though that has some issues to work through still: #1449 #1290).

Yes the password was changed to prevent accidental sync from the old device. But the migration didn't carry out carefully in the ivory coast as users from some of the site used to setup the new device and started sync before they sync all pending data from the old device and uninstall the mSupply app. Then again they requested to sync the old pending data from the old device in which case we had to revert/update sync password from time to time.

Also its a bit hard to track device hardware id and to monitor which device should we give access to sync or for which not to specially in the situation like this where the remoting to the mobile doesn't work properly.

**Comment by joshxg** [comment]
Sounds like a settings page is agreed, for now.

I think one more important detail:
@ujwalSussol @raviSussol do you have any ideas on how to prevent a standard user from accessing the settings page and changing these details? I think avoiding a hardcoded password is important.

Ideas so far:
- Sync users with a "Admin role"  [introduce permissions/per use customizations to mobile]
- Send a 'admin password' with initial sync dump [just another password to manage...]

**Comment by Chris-Petty** [comment]
I agree generally that we should do the settings page too, so may as well go ahead for this. My crudest idea for securing it is to sync the 'sussol' user. When the tablet URL/Pass isn't working, it won't try authenticate on the server and won't fail haha, using the local deets instead. So long as there is an issue to do it better, ultimately with permission's I think which should have some broader app thinking put into it.

**Comment by ujwalSussol** [comment]
> Sounds like a settings page is agreed, for now.
> 
> I think one more important detail:
> @ujwalSussol @raviSussol do you have any ideas on how to prevent a standard user from accessing the settings page and changing these details? I think avoiding a hardcoded password is important.
> 
> Ideas so far:
> 
> * Sync users with a "Admin role"  [introduce permissions/per use customizations to mobile]
> * Send a 'admin password' with initial sync dump [just another password to manage...]

@joshxg : I don't think we need to worry about users getting access to setting page and changing things.  If they do, the sync will simply not work.  The sync user name should not be changeable.

**Comment by ujwalSussol** [comment]
@joshxg  :  having said the above, I think "Sussol" user or any user with admin right can have access to the setting page.

**Comment by joshxg** [comment]
@ujwalSussol no problem - since we don't have permissions on mobile, will use the `Sussol` user for now. I think it's a bit too hacky personally - will make a new issue to potentially change as chris suggests

**Comment by ujwalSussol** [comment]
> @ujwalSussol no problem - since we don't have permissions on mobile, will use the `Sussol` user for now. I think it's a bit too hacky personally - will make a new issue to potentially change as chris suggests

👍   Great ... the setting page will help a lot in the future for guys who are having to do front line support.

**Comment by joshxg** [comment]
Spoke more with @ujwalSussol :

Probably not a good idea to change `sussol` user being able to login to every site when it's been decided that it's a bit too much of a security problem. 

Thinking: Just send a boolean `isAdmin: true/false` in the body of a request to login - can save this in the schema as `isAdmin` for offline.

Don't have to sync users/permissions etc. It is a precedent for permissions though

**Comment by raviSussol** [comment]
@Chris-Petty @joshxg Sounds like a good idea! 👍

**Comment by adrianwb** [comment]
I like the idea of a mobile settings page and, as I understand it, the main issue here is how to secure that page. 

Currently, on initialisation of a desktop site, the users with login rights to the stores active on that site get synced from the central sync server, but for mobile, no users, permissions or prefs currently get synced. Also on desktop, changing the sync password/URL can only be done by an admin user who has access to the prefs on _both_ sides _and_ knows the sync auth code => we manually change the sync password on the central sync server _and_ again manually on the remote site - no passwords get synced. I'm assuming that we don't want end users to be changing sync settings without explicit direction from us, and only after we've changed the corresponding settings on the central sync server? 

=> what about using the tablet's hardware UUID (or even just the first XXX characters of it) as the password for the local mobile admin user to access the settings, but keep the sync password/URL behind an additional auth code that only we know (that could be fixed as it is now, or it could be based on the hardware UUID or site name, but using a pattern that is not so obvious to the local user)?

I don't see much benefit in syncing users and passwords from the central sync server - apart from the initial setup, we've managed to avoid that for the most part, even on desktop. For mobile to be able to decrypt user passwords sent from the central server, it would either need to have the same private key, or we'd need to re-encrypt them with another key that mobile does have access to (or can work out) and that just ends up making things way more complicated. I'm not sure that we need any changes on the server side at all?

**Comment by joshxg** [comment]
Thanks @adrianwb 

- Use hardware UUID as the password is definitely the simplest option. A little bit unstable given androids need to keep changing the API (some android versions 8+ the UUID is changing.. some 
OS's the ID isn't accessible from anywhere a normal user can access). Also not too secure. If we keep the sync details behind another password - the algorithm or hardcoded password would be in an open source repo, or add annoyances to building.

- Sending a `isAdmin` boolean in the body of a login response from the server is simple also? We can then store that on mobile - then the only user able to access to change the sync url/password is one who can change these on desktop?

**Comment by adrianwb** [comment]
> 
> 
> Thanks @adrianwb
> 
>     * Use hardware UUID as the password is definitely the simplest option. A little bit unstable given androids need to keep changing the API (some android versions 8+ the UUID is changing.. some
>       OS's the ID isn't accessible from anywhere a normal user can access, would need to get it from the server). Also not too secure. If we keep the sync details behind another password - the algorithm or hardcoded password would be in an open source repo, or add annoyances to building.

I know that there have been recent changes on desktop to send the hardware UUID as part of the sync header so that we can block sync if that changes (i.e. if it's coming from another machine). I'm guessing that hasn't made its way to mobile yet, in which case implementing it could cause problems when the OS is updated.

In this case, I think the intention is just to make it difficult/unlikely that any regular user will change the settings rather than completely secure, but your point about storing any kind of password-checking algorithm or hard-coded auth code in an open source repo is something that we will have to deal with in the not-too distant future with omSupply...

>     * Sending a `isAdmin` boolean in the body of a login response from the server is simple also? We can then store that on mobile - then the only user able to access to change the sync url/password is one who can change these on desktop?

Ah, I get it now - yes, that sounds like it'll work without too many changes server-side.

**Comment by joshxg** [comment]
Closed by #1522

**Comment by Rebecca367** [comment]
## Testing

#### Setup: Will need to be running a server with the linked mSupply code above, in order for the server to send `isAdmin` header.

- [x] Login with a user WITH `view/edit preferences` permission. There should be a `Settings` button in the bottom right of the menu page.
- [Failed] Login with a user WITHOUT `view/edit preferences` permission. There should be NO `Settings` button in the bottom right of the menu page.
- [x] Clicking the SETTINGS button navigates to a SETTINGS page

**Note: Can verify details by using the Realm Explorer and entering 'Setting' in the object string**
#### SETTINGS PAGE:
- [x] SYNC URL
    - [x] Displays the current sync url correctly
    - [x] Clicking to edit opens a modal, whose title gives a warning
    - [x] Opening the modal has the sync url entered already
    - [x] You CANNOT enter an INVALID URL as defined by [RFC 3986](https://tools.ietf.org
/html/rfc3986) (Just try some whacky ones)
    - [x] When entering a valid URI and clicking OK, is saved.
- [x] SYNC PASSWORD
    - [x] Does not show the current password ever
    - [x] uses password hiding when typing
    - [x] Entering a new password saves it 
- [x] REALM EXPLORER
    - [x] Navigates to the Realm Explorer page
- [x] EXPORT DATA
    - [x] Exports database (to downloads folder)

There is setting icon with a user WITHOUT `view/edit preferences` permission
<img width="1440" alt="Screen Shot 2019-12-11 at 8 32 51 AM" src="https://user-images.githubusercontent.com/57238075/70563202-b02d0f00-1bf2-11ea-9da0-621cdb8d99c4.png">
<img width="1440" alt="Screen Shot 2019-12-11 at 8 34 14 AM" src="https://user-images.githubusercontent.com/57238075/70563214-b58a5980-1bf2-11ea-98bd-a1d58c4e885f.png">

**Comment by emilyporrello** [comment]
Docs updated

Settings button explained in Getting Started page (with updated screenshots of main screen); details outlined in new Mobile Settings page
https://docs.msupply.foundation/en:mobile:user_guide:getting_started
https://docs.msupply.foundation/en:mobile:user_guide:mobile_settings
